wallet2/db/db.go

```diff
@@ -0,0 +1,549 @@
+	TransactionTable = "transactions"
+	PairTable        = "pairs"
+)
+
+type DB interface {
+	CreateTables() error
+
+	InsertIntoAddress(addr *Address) (*Address, error)
+	InsertIntoTransaction(t *Transaction) (*Transaction, error)
+	InsertIntoPair(key string, value string) (*Pair, error)
+
+	UpdateAddressLabel(addr *Address) (*Address, error)
+
+	GetAddressByID(id int) (*Address, error)
+	GetAddressByAddress(address string) (*Address, error)
+	GetAddressByPath(p string) (*Address, error)
+
+	GetTransactionByID(id int) (*Transaction, error)
+	GetTransactionByTxID(id string) (*Transaction, error)
+
+	GetPairByKey(key string) (*Pair, error)
+	GetTotalRecords(tableName string, query string, args ...any) (int64, error)
+
+	GetAllAddresses() ([]Address, error)
+	GetAllAddressesWithTotalRecords(pageIndex, pageSize int) ([]Address, int64, error)
+
+	GetAllTransactions(query string, args ...any) ([]Transaction, error)
+	GetAllTransactionsWithTotalRecords(pageIndex, pageSize int, query string, args ...any) ([]Transaction, int64, error)
+}
+
+type db struct {
+	*sql.DB
+	context.Context
+}
+
+type Address struct {
+	ID         int       `json:"id"`          // id of wallet
+	Address    string    `json:"address"`     // Address in the wallet
+	PublicKey  string    `json:"public_key"`  // Public key associated with the address
+	Label      string    `json:"label"`       // Label for the address
+	Path       string    `json:"path"`        // Path for the address
+	IsImported bool      `json:"is_imported"` // imported for purpose
+	CreatedAt  time.Time `json:"created_at"`
+}
+
+type Transaction struct {
+	ID          int       `json:"id"`
+	TxID        string    `json:"tx_id"`
+	Address     string    `json:"address"`
+	BlockHeight uint32    `json:"block_height"`
+	BlockTime   uint32    `json:"block_time"`
+	PayloadType string    `json:"payload_type"`
+	Data        string    `json:"data"`
+	Description string    `json:"description"`
+	Amount      int64     `json:"amount"`
+	Status      int       `json:"status"`
+	CreatedAt   time.Time `json:"created_at"`
+}
+
+type Pair struct {
+	Key       string
+	Value     string
+	CreatedAt time.Time
+}
+
+func NewDB(ctx context.Context, path string) (DB, error) {
+	dbInstance, err := sql.Open("sqlite", path)
+	if err != nil {
+		return nil, ErrCouldNotOpenDatabase
+	}
+
+	return &db{
+		dbInstance,
+		ctx,
+	}, nil
+}
+
+func (d *db) CreateTables() error {
+	if err := d.createAddressTable(); err != nil {
+		return err
+	}
+
+	if err := d.createTransactionTable(); err != nil {
+		return err
+	}
+
+	return d.createPairTable()
+}
+
+func (d *db) createAddressTable() error {
+	addressQuery := fmt.Sprintf("CREATE TABLE %s (id INTEGER PRIMARY KEY AUTOINCREMENT,"+
+		" address VARCHAR, public_key VARCHAR, label VARCHAR, path VARCHAR, is_imported BOOLEAN, created_at TIMESTAMP)",
+		AddressTable)
+
+	_, err := d.ExecContext(d.Context, addressQuery)
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
```

Comment:
This code is not lovely! Any other way to check if a table exists?

Reply:
I've searched for alternatives, but unfortunately, it seems there's no solution except for this one